File: src/main/java/org/pacien/pandoc/filter/plantuml/Filter.java
package org.pacien.pandoc.filter.plantuml;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.TextNode;
import net.sourceforge.plantuml.FileFormat;
import net.sourceforge.plantuml.FileFormatOption;
import net.sourceforge.plantuml.SourceStringReader;

import java.io.*;
import java.util.Iterator;
import java.util.stream.Collectors;

final public class Filter {

  private static final String BEGIN_TAG = "\\begin{tikzpicture}[yscale=-1]";
  private static final String LINE_SEP = "\n";
  private static final String TYPE_KEY = "t";
  private static final String CONTENT_KEY = "c";
  private static final String CODE_BLOCK_TYPE = "CodeBlock";
  private static final String RAW_BLOCK_TYPE = "RawBlock";
  private static final String PLANTUML_TYPE = "puml";
  private static final String LATEX_TYPE = "latex";
  private static final int META_INDEX = 0;
  private static final int META_PROP_INDEX = 1;
  private static final int META_PROP_TYPE_INDEX = 0;
  private static final int CONTENT_INDEX = 1;

  private static String plantumlToLatex(String puml) throws IOException {
    try (ByteArrayOutputStream s = new ByteArrayOutputStream()) {
      new SourceStringReader(puml).generateImage(s, new FileFormatOption(FileFormat.LATEX_NO_PREAMBLE));
      try (BufferedReader r = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(s.toByteArray())))) {
        return BEGIN_TAG + LINE_SEP + r.lines().filter(l -> !l.equals(BEGIN_TAG)).collect(Collectors.joining(LINE_SEP));
      }
    }
  }

  private static void renderPlantumlNode(ObjectNode n) throws IOException {
    String puml = n.get(CONTENT_KEY).get(CONTENT_INDEX).asText();
    String tikz = plantumlToLatex(puml);

    n.set(TYPE_KEY, TextNode.valueOf(RAW_BLOCK_TYPE));
    ((ArrayNode) n.get(CONTENT_KEY)).removeAll()
    .add(TextNode.valueOf(LATEX_TYPE))
    .add(TextNode.valueOf(tikz));
  }

  private static boolean isPlantumlNode(JsonNode n) {
    return n.path(TYPE_KEY).asText().equals(CODE_BLOCK_TYPE) &&
           n.path(CONTENT_KEY).path(META_INDEX).path(META_PROP_INDEX).path(META_PROP_TYPE_INDEX).asText().equals(PLANTUML_TYPE);
  }

  private static void walk(JsonNode n) throws IOException {
    if (isPlantumlNode(n))
      renderPlantumlNode((ObjectNode) n);
    else if (n.isContainerNode())
      for (Iterator<JsonNode> i = n.elements(); i.hasNext(); ) walk(i.next());
  }

  public static void filter(InputStream i, OutputStream o) throws IOException {
    ObjectMapper m = new ObjectMapper();
    JsonNode t = m.readTree(i);
    if (t != null) {
      walk(t);
      m.writeValue(o, t);
    }
  }

  public static void main(String args[]) {
    try {
      filter(System.in, System.out);
    } catch (IOException e) {
      e.printStackTrace();
      System.exit(1);
    }
  }

  private Filter() {
    // static class
  }

}

File: src/main/kotlin/org/pacien/pandoc/filter/plantuml/Filter.kt
/*
 * pandoc-filter-plantuml, a Pandoc AST filter rendering PlantUML diagrams
 * Copyright (C) 2018 Pacien TRAN-GIRARD
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.pacien.pandoc.filter.plantuml

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.node.TextNode
import java.io.InputStream
import java.io.OutputStream

object Filter {
  private val mapper = ObjectMapper()

  private inline fun <T> T.conditionally(condition: Boolean, block: (T) -> T): T =
    if (condition) block(this) else this

  private inline fun <T, A> T.withNonNull(value: A?, block: (T, A) -> T): T =
    if (value != null) block(this, value) else this

  private fun JsonNode.isCodeBlock() = type() == "CodeBlock"
  private fun JsonNode.isPlantUmlBlock() = isCodeBlock() && "puml" in classNames()

  private fun Latex.resizeBox(attrs: Map<String, String>) =
    resizeBox(attrs["width"] ?: "!", attrs["height"] ?: "!")

  private fun Latex.setOptions(classes: List<String>, attrs: Map<String, String>) =
    this
      .conditionally("width" in attrs || "height" in attrs) { it -> it.resizeBox(attrs) }
      .conditionally("centered" in classes, Latex::centering)
      .withNonNull(attrs["caption"], Latex::caption)
      .withNonNull(attrs["label"], Latex::label)
      .conditionally("caption" in attrs || "label" in attrs, Latex::figure)

  private fun arrayNodeOf(type: String, content: String): ArrayNode =
    mapper.createArrayNode()
      .add(TextNode.valueOf(type))
      .add(TextNode.valueOf(content))

  private fun renderPlantumlNode(node: ObjectNode) {
    val puml = node.content()
    val tikz = PlantUml.renderTikz(puml)
    val block = tikz.setOptions(node.classNames(), node.attributeMap())
    node.setBlock("RawBlock", arrayNodeOf("latex", block.raw()))
  }

  private fun walk(node: JsonNode): Unit = when {
    node.isPlantUmlBlock() -> renderPlantumlNode(node as ObjectNode)
    else -> node.forEach(Filter::walk)
  }

  fun filter(input: InputStream, output: OutputStream) {
    mapper.readTree(input)?.let { tree ->
      walk(tree)
      mapper.writeValue(output, tree)
    }
  }
}
